### PR TITLE
ARROW-7020: [Java] Fix the bugs when calculating vector hash code

### PR DESCRIPTION
**Pull request by liyafan82**
When calculating the hash code for a value in the vector, the validity bit must be taken into account.

### CODE REVIEW

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 1]
```diff
@@ -693,6 +693,9 @@ public int hashCode(int index) {
 
     @Override
     public int hashCode(int index, ArrowBufHasher hasher) {
+      if (isNull(index)) {
```
Comment (emkornfield) [inline java/vector/src/main/codegen/templates/UnionVector.java]: I don't think this is needed given the current implementation, is index would point to ZeroVector which would hash?

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java [hunk 2]
```diff
@@ -1373,6 +1373,9 @@ public int hashCode(int index) {
 
   @Override
   public int hashCode(int index, ArrowBufHasher hasher) {
+    if (isNull(index)) {
+      return 0;
```
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: is zero really a good value to return for hashing purposes?

File: java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java [hunk 3]
```diff
@@ -2642,6 +2642,26 @@ public void testEqualsWithIndexOutOfRange() {
     }
   }
 
+  @Test
+  public void testNullHashCode() {
```
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java]: please separate this to a test per type, it is also missing a test for union types?

File: java/vector/src/main/codegen/templates/UnionVector.java (hunk 1 above)
Comment (liyafan82) [reply java/vector/src/main/codegen/templates/UnionVector.java]: I think you are right. 

However, that means there is some inconsistency in the current implementation. Maybe it is another problem that needs to be fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java (hunk 2 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: According to my experience, zero is commonly used as hash code for nulls.

File: java/vector/src/main/codegen/templates/UnionVector.java (hunk 1 above)
Comment (liyafan82) [reply java/vector/src/main/codegen/templates/UnionVector.java]: There is another problem here: the result of getVector(index) can be null. I have fixed this problem, and the above problem, by revising the implementation of isNull(index). 

Pls take a look.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java (hunk 2 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: I also find this in "Effective Java, 3rd ed." pp. 51:

If the value of the field is null, use 0 (or some other constant, but 0 is traditional).

File: java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java (hunk 3 above)
Comment (liyafan82) [reply java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java]: Good suggestion. Thanks.
I have added separate test cases for FixedWidthVector, VariableWidthVector and UnionVector.

**Comment by emkornfield** [comment]
+1, thank you.

**Pull request closed without merging**